services/ingestion/sealing_verifier.go

```diff
@@ -0,0 +1,256 @@
+)
+
+var _ models.Engine = &SealingVerifier{}
+
+// SealingVerifier verifies that soft finality events received over the Access polling API match the
+// actually sealed results from the event stream.
+type SealingVerifier struct {
+	*models.EngineStatus
+
+	logger zerolog.Logger
+
+	client *requester.CrossSporkClient
+	chain  flowGo.ChainID
+
+	eventsHash     *pebble.EventsHash
+	blocksToVerify map[uint64]flow.Identifier
+
+	mu sync.Mutex
+}
+
+// NewSealingVerifier creates a new sealing verifier.
+func NewSealingVerifier(
+	logger zerolog.Logger,
+	client *requester.CrossSporkClient,
+	chain flowGo.ChainID,
+	eventsHash *pebble.EventsHash,
+) *SealingVerifier {
+	return &SealingVerifier{
+		EngineStatus: models.NewEngineStatus(),
+		logger:       logger,
+		client:       client,
+		chain:        chain,
+		eventsHash:   eventsHash,
+	}
+}
+
+// Stop the engine.
+func (v *SealingVerifier) Stop() {
+	v.MarkDone()
+	<-v.Stopped()
+}
+
+// AddBlock adds a block to the sealing verifier for verification when the sealed data is received.
+func (v *SealingVerifier) AddBlock(events flow.BlockEvents) error {
+	hash, err := CalculateHash(events)
+	if err != nil {
+		return fmt.Errorf("failed to calculate hash for block %d: %w", events.Height, err)
+	}
+
+	if err := v.eventsHash.Store(events.Height, hash); err != nil {
+		return fmt.Errorf("failed to store events hash for block %d: %w", events.Height, err)
+	}
+
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	v.blocksToVerify[events.Height] = hash
+
+	return nil
+}
+
+// Run executes the sealing verifier.
+// This method will block until the context is canceled or an error occurs.
+func (v *SealingVerifier) Run(ctx context.Context) error {
+	defer v.MarkStopped()
+
+	lastVerifiedHeight, err := v.eventsHash.ProcessedSealedHeight()
+	if err != nil {
+		return fmt.Errorf("failed to get processed sealed height: %w", err)
+	}
+
+	var eventsChan <-chan flow.BlockEvents
+	var errChan <-chan error
+
+	subscriptionCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	lastReceivedHeight := lastVerifiedHeight + 1
+	connect := func(height uint64) error {
+		var err error
+		eventsChan, errChan, err = v.client.SubscribeEventsByBlockHeight(
+			subscriptionCtx,
+			height,
+			blocksFilter(v.chain),
+			access.WithHeartbeatInterval(1),
+		)
+		return err
+	}
+
+	if err := connect(lastReceivedHeight); err != nil {
+		return fmt.Errorf("failed to subscribe for finalized block events on height: %d, with: %w", lastReceivedHeight, err)
+	}
+
+	v.MarkReady()
+	for {
+		select {
+		case <-ctx.Done():
+			v.logger.Info().Msg("sealing verifier received done signal")
+			return nil
+
+		case <-v.Done():
+			v.logger.Info().Msg("sealing verifier received stop signal")
+			cancel()
+			return nil
+
+		case blockEvents, ok := <-eventsChan:
+			if !ok {
+				if ctx.Err() != nil {
+					return nil
+				}
+				return fmt.Errorf("failed to receive block events: %w", err)
+			}
+
+			if err := v.verifyBlock(blockEvents); err != nil {
```

Comment:
can this race with finalized somehow ? ( I mean if finalized is interrupt and reconnect etc ) and sealed passes finalized height ?

Reply:
yea, you're right. [c200c57](https://github.com/onflow/flow-evm-gateway/pull/757/commits/c200c574e743df1ed94c1ca9ae80d9ac4ebb976c)